rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,329 @@
+		// read before.
+		newBytesRead, err := IOMgrConn.Conn.Read(prefixBuf[bytesRead : bytesRead+1])
+		if newBytesRead == 0 {
+			if io.EOF == err {
+				return bytesRead, n, nil
+			} else if err != nil {
+				return bytesRead, n, err
+			}
+			// A Reader should not return (0, nil), but if it does,
+			// it should be treated as no-op (according to the
+			// Reader contract). So let's go on...
+			continue
+		}
+		bytesRead += newBytesRead
+		// Now present everything read so far to the varint decoder and
+		// see if a varint can be decoded already.
+		n, varIntBytes = proto.DecodeVarint(prefixBuf[:bytesRead])
+	}
+
+	return bytesRead, n, err
+}
+
+func (IOMgrConn IOMgrConnection) writeMessage(messageTypeID uint64, m proto.Message) (int, error) {
+
+	// Function writeMessage()
+	// source: https://github.com/stashed/stash/blob/master/vendor/github.com/matttproud/golang_protobuf_extensions/pbutil/encode.go
+	//
+	// Copyright 2013 Matt T. Proud
+	//
+	// Licensed under the Apache License, Version 2.0 (the "License");
+	// you may not use this file except in compliance with the License.
+	// You may obtain a copy of the License at
+	//
+	//     http://www.apache.org/licenses/LICENSE-2.0
+	//
+	// Unless required by applicable law or agreed to in writing, software
+	// distributed under the License is distributed on an "AS IS" BASIS,
+	// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	// See the License for the specific language governing permissions and
+	// limitations under the License.
+
+	var buf [binary.MaxVarintLen32]byte
+	encodedLength := binary.PutUvarint(buf[:], messageTypeID)
+	IOMgrConn.Conn.Write(buf[:encodedLength])
+
+	em, err := proto.Marshal(m)
+	if nil != err {
+		return 0, err
+	}
+
+	encodedLength = binary.PutUvarint(buf[:], uint64(proto.Size(m)))
+
+	vib, err := IOMgrConn.Conn.Write(buf[:encodedLength])
+	if nil != err {
+		return vib, err
+	}
+
+	pbmmb, err := IOMgrConn.Conn.Write(em)
+
+	return vib + pbmmb, err
+}
+
+func (IOMgrConn IOMgrConnection) readMessage(messageTypeID uint64, m proto.Message) (n int, err error) {
+
+	// Function readMessage()
+	// source: https://github.com/stashed/stash/blob/master/vendor/github.com/matttproud/golang_protobuf_extensions/pbutil/decode.go
+	//
+	// Copyright 2013 Matt T. Proud
+	//
+	// Licensed under the Apache License, Version 2.0 (the "License");
+	// you may not use this file except in compliance with the License.
+	// You may obtain a copy of the License at
+	//
+	//     http://www.apache.org/licenses/LICENSE-2.0
+	//
+	// Unless required by applicable law or agreed to in writing, software
+	// distributed under the License is distributed on an "AS IS" BASIS,
+	// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	// See the License for the specific language governing permissions and
+	// limitations under the License.
+
+	var prefixBuf [binary.MaxVarintLen64]byte
+	var bytesRead, varIntBytes int
+	var messageLength uint64
```

Comment:
messge length is 32 bit

Reply:
@cxxman  The message length is decoded by the function `proto.DecodeVarint` which returns a uint64.
Does it really worst convert that length to 32bits and add complexity to the code and extra cycles?

Reply:
I'd say this is sort of micro-optimization in favor of not following the required communication protocol. Reading varint32 does not add any extra cycles, because varint32 is finally shorter that 64. Binary encoding is such that if you got wrong varint32 you stop at 5th byte, but if you assume to read varint64 it can be treated as valid varint64 which is up to 10 bytes. Also you don't validate it. ANY length-like parameter must be validated.